tests/irdl/test_declarative_assembly_format.py

```diff
@@ -1678,11 +1678,11 @@ def constr(
 
     @irdl_op_definition
     class TwoOperandsNestedVarOp(IRDLOperation):
-        T = Annotated[Attribute, ConstraintVar("T")]
+        T: ClassVar[VarConstraint[Attribute]] = VarConstraint("T", AnyAttr())
 
         name = "test.two_operands_one_result_with_var"
         res = result_def(T)
-        lhs = operand_def(ParamOne[T])
+        lhs = operand_def(ParamOne[Attribute].constr(p=T))
```

Comment:
I do have to say, I really dislike the verbosity of this, it will make it harder for newcomers to get started I think.

Reply:
Yeah it definitely got less pretty. One potential alternative is `ParamOneConstr(p=T)`, would you like me to try that instead? It's one more thing to import, but potentially worth the effort.

Reply:
I just tested it locally and that API works and makes Pyright happy, it really doesn't want you to omit the type parameter even if it could be inferred in this case. I'm still inclined towards this more verbose API for the time being, with the hope that Pyright gets smart enough at some point to drop the unnecessary type parameter. We could also file it as a bug.

Reply:
https://github.com/microsoft/pyright/issues/9190

Reply:
What about making `constr` a staticmethod instead? I just tried it locally and my pyright at least seems happy?

Reply:
This allows you to do `ParamOne.constr(p=T)` instead of `ParamOne[Attribute].constr(p=T)`

Reply:
it doesn't work for me

Reply:
How do you feel about iterating on this after this is merged and Pyright is updated?

Reply:
Fine with me

Reply:
https://peps.python.org/pep-0696/ also potentially relevant for making this api nicer

Reply:
Yep, I look forward to being able to use the new, shiny APIs. I'm not sure when it would be wise to drop support for older Pythons, according to this calendar 3.10 is supported for another couple of years: https://endoflife.date/python